File: web/src/api/xunyin/city.ts
import request from '@/utils/request'

export interface City {
  id: string
  name: string
  province: string
  latitude: number
  longitude: number
  iconAsset?: string
  coverImage?: string
  description?: string
  explorerCount: number
  journeyCount?: number
  bgmUrl?: string
  orderNum: number
  status: string
  createTime: string
  updateTime: string
}

export interface CityQuery {
  name?: string
  province?: string
  status?: string
  pageNum?: number
  pageSize?: number
}

export interface CityForm {
  id?: string
  name: string
  province: string
  latitude: number
  longitude: number
  iconAsset?: string
  coverImage?: string
  description?: string
  bgmUrl?: string
  orderNum?: number
  status?: string
}

export function listCity(query: CityQuery) {
  return request<{ data: { list: City[]; total: number; pageNum: number; pageSize: number } }>({
    url: '/admin/cities',
    method: 'get',
    params: query,
  }).then((res: any) => res.data)
}

export function getCity(id: string) {
  return request<{ data: City }>({
    url: `/admin/cities/${id}`,
    method: 'get',
  }).then((res: any) => res.data)
}

export function addCity(data: CityForm) {
  return request({
    url: '/admin/cities',
    method: 'post',
    data,
  })
}

export function updateCity(data: CityForm) {
  return request({
    url: `/admin/cities/${data.id}`,
    method: 'put',
    data,
  })
}

export function delCity(id: string) {
  return request({
    url: `/admin/cities/${id}`,
    method: 'delete',
  })
}

export function updateCityStatus(id: string, status: string) {
  return request({
    url: `/admin/cities/${id}/status`,
    method: 'patch',
    data: { status },
  })
}

export function exportCities(query: CityQuery) {
  return request<Blob>({
    url: '/admin/cities/export',
    method: 'get',
    params: query,
    responseType: 'blob',
  })
}

File: app_flutter/lib/models/city.dart
/// 城市模型
class City {
  final String id;
  final String name;
  final String province;
  final double latitude;
  final double longitude;
  final String? iconAsset;
  final String? coverImage;
  final String? description;
  final int explorerCount;
  final String? bgmUrl;
  final int? journeyCount;

  City({
    required this.id,
    required this.name,
    required this.province,
    required this.latitude,
    required this.longitude,
    this.iconAsset,
    this.coverImage,
    this.description,
    this.explorerCount = 0,
    this.bgmUrl,
    this.journeyCount,
  });

  factory City.fromJson(Map<String, dynamic> json) {
    final coverImage = json['coverImage'] as String?;
    final iconAsset = json['iconAsset'] as String?;
    final bgmUrl = json['bgmUrl'] as String?;
    return City(
      id: json['id'] as String,
      name: json['name'] as String,
      province: json['province'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      iconAsset: iconAsset?.isNotEmpty == true ? iconAsset : null,
      coverImage: coverImage?.isNotEmpty == true ? coverImage : null,
      description: json['description'] as String?,
      explorerCount: json['explorerCount'] as int? ?? 0,
      bgmUrl: bgmUrl?.isNotEmpty == true ? bgmUrl : null,
      journeyCount: json['journeyCount'] as int?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'province': province,
    'latitude': latitude,
    'longitude': longitude,
    'iconAsset': iconAsset,
    'coverImage': coverImage,
    'description': description,
    'explorerCount': explorerCount,
    'bgmUrl': bgmUrl,
    'journeyCount': journeyCount,
  };
}

/// 附近城市（带距离）
class NearbyCity extends City {
  final double distance;

  NearbyCity({
    required super.id,
    required super.name,
    required super.province,
    required super.latitude,
    required super.longitude,
    super.iconAsset,
    super.coverImage,
    super.description,
    super.explorerCount,
    super.bgmUrl,
    super.journeyCount,
    required this.distance,
  });

  factory NearbyCity.fromJson(Map<String, dynamic> json) {
    final coverImage = json['coverImage'] as String?;
    final iconAsset = json['iconAsset'] as String?;
    final bgmUrl = json['bgmUrl'] as String?;
    return NearbyCity(
      id: json['id'] as String,
      name: json['name'] as String,
      province: json['province'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      iconAsset: iconAsset?.isNotEmpty == true ? iconAsset : null,
      coverImage: coverImage?.isNotEmpty == true ? coverImage : null,
      description: json['description'] as String?,
      explorerCount: json['explorerCount'] as int? ?? 0,
      bgmUrl: bgmUrl?.isNotEmpty == true ? bgmUrl : null,
      journeyCount: json['journeyCount'] as int?,
      distance: (json['distance'] as num).toDouble(),
    );
  }
}
